### DISCUSSION TITLE
aioble aysnc await not allowing capture of unsolicited events

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#13769] [Libraries & Drivers]
I am not sure how to get around this problem as I need to catch an event that may not occur. The `async await` is usually very convenient as you `await` for an event, but that assumes you know that it is coming. I don't want to `await` and event that might not come.

I have a situation where I am writing a CCCD to a characteristic that is secured and I am not bonded. There are many possibilities. 
One is that I get an insufficient authentication error. I can trap that GattError and initiate pairing and repeat the operation once bonded.
Second is that the library initiates pairing. The btstack does not, so I don't have to worry about it.
Third is that the device can issue a security request which the btstack does handle but I (and aioble) have no idea that it happened, It appears that the device finishes the operation as I get no error and I do not have to repeat the operation.

I do not want to `await` such an event as it might never happen. Some playing around with tasks maybe?
Looking for ideas.

**Comment by rkompass** [comment]
In which context do you want to use the `async await`?
The async `wait_for()` has a timout argument. Would this help?
From my naive understanding the event not occurring would be translated to a certain timeout having passed.

**Reply by brianreinhold** [reply]
Timeouts are always troubling in communications. You pick a value which may be too long for some devices and too short for others. You need to be signaled when something is done. Right now a timeout is exactly what I am doing. I have code that looks like this:
```
await self.msmt_char.subscribe(False, True)
await check_if_security_request_sent(self.phd_info, self.deviceConnection)
```
I am using the `aioble subscribe()` method on its `ClientCharacteristic` to set the CCCD to indicate in this case. However, the descriptor is secured and what the peer device does is to respond with a security request. Now it could have sent an `insufficient authentication` error instead. But because it sends the security request, the `btstack` handles it and invokes pairing. The bonding completes successfully and the peer responds to the CCCD write. `aioble` and my application have no idea the bonding took place.

So what I have done is to add a method to aioble that does everything the pairing method does except request pairing. But it does register for the encryption change event. So by calling this method if no security request is sent, 3 seconds will elapse and it will time out. But that's a horrible slow down if you have to wait 3 seconds on every client action since you have no ideas what devices will pair and on which action the security will sit if it does pair, and if they will even choose to use a security request.

There has got to be a better way where I can be notified that pairing happened and have aioble notified as well.  There might be a way to do it with tasks but I am still pretty much a greenhorn in this arena!

**Reply by brianreinhold** [reply]
What I ended up doing is creating a task which ran at the start of the connection, The task was passed the deviceConnection and phd_info objects. The task called that new aioble method I added to trigger on an encryption update at which point aioble would have the information and so would my application. The task would also terminate. Otherwise there was a long timeout or task cancellation in case of no unsolicited pairing actions. Seems to work.

**Reply by rkompass** [reply]
Nice. Congratulations.
I hope you will share your problem descriptions and your code (aioble addition and your code making use of it) at some point.
Do you plan to do a PR?
I'm no Bluetooth expert at all. But I have the feeling this may be relevant for me in the future.

**Reply by brianreinhold** [reply]
@rkompass  What is a PR? I could write up what we have done and what we still need to do. For the most part (with some hacks) we are able to pair and do bonded reconnects (encryption) in when both security requests are issued (handled by the btstack) or authentication failures are send (handled by the application). Wish both were handled by the application - would greatly reduce complexity.

**Reply by rkompass** [reply]
PR refers to PullRequest and assumes you are familiar with Git. 

>  I could write up what we have done and what we still need to do.

That would be nice. It would correspond to the spirit of free software and it would help others to benefit from you insights/solutions.
As I'm not the Github expert myself, I think posting your problems/solutions with some code examples here in discussions would be fine. Perhaps as a new topic with descriptive heading.

**Comment by brianreinhold** [comment]
Okay, I know what a pull request is, but the work done to solve this problem is not complete enough to be added to this project. In addition, there are issues that are being worked around which should not be the case in the final distribution.

I suppose the biggest case in point is that sometimes the btstack handles the pairing and encryption and aioble and the application (my code) knows nothing about it. This happens when the device issues a security request. If the device does not issue security requests, my code has to initiate pairing via aioble. The correct solution would do one or the other - the btstack handles all pairing/bonding/encryption on reconnect needs, or the application does. I would prefer the latter!

RIght now I handle the security request by making a task that registers for the encryption changed event in aioble which is started at connection. To do that I had to add another method to aioble does exactly what the pairing method does except call pair! When the pairing invoked by the btstack is done, aioble gets signaled which signlas my app. Both aioble and my app now know that pairing/bonding/encryption occurred.

However, if my app needs to initiate pairing (needed when the device does not do security requests) I have to kill the security request task or it will get the event and my pair call will 'await' on the pair method until it times out.

As you can see the solution is not ideal! But it works.
